pkg/registry/core/service/storage/rest.go

```diff
@@ -675,14 +676,31 @@ func initClusterIP(service *api.Service, allocator ipallocator.Interface) (bool,
 			return false, errors.NewInternalError(fmt.Errorf("failed to allocate a serviceIP: %v", err))
 		}
 		service.Spec.ClusterIP = ip.String()
+		// default the IPFamily appropriately when the gate is on
+		if utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) && service.Spec.IPFamily == nil {
+			ipFamily := api.IPv4Protocol
+			if netutil.IsIPv6(ip) {
+				ipFamily = api.IPv6Protocol
+			}
+			service.Spec.IPFamily = &ipFamily
+		}
 		return true, nil
 	case service.Spec.ClusterIP != api.ClusterIPNone && service.Spec.ClusterIP != "":
 		// Try to respect the requested IP.
-		if err := allocator.Allocate(net.ParseIP(service.Spec.ClusterIP)); err != nil {
+		ip := net.ParseIP(service.Spec.ClusterIP)
+		if err := allocator.Allocate(ip); err != nil {
 			// TODO: when validation becomes versioned, this gets more complicated.
 			el := field.ErrorList{field.Invalid(field.NewPath("spec", "clusterIP"), service.Spec.ClusterIP, err.Error())}
 			return false, errors.NewInvalid(api.Kind("Service"), service.Name, el)
 		}
+		// default the IPFamily appropriately when the gate is on
+		if utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) && service.Spec.IPFamily == nil {
```

Comment:
could move this outside the `switch` rather than having the same code in both branches (putting the `return false, nil` case into a `default` branch)

Reply:
yeah, agree